databricks_cli/pipelines/cli.py

```diff
@@ -57,30 +63,97 @@ def deploy_cli(api_client, spec_arg, spec):
 
 
 @click.command(context_settings=CONTEXT_SETTINGS,
-               short_help='Stops a delta pipeline and cleans '
-                          'up Databricks resources associated with it')
+               short_help='Stops a delta pipeline and deletes its associated Databricks resources')
 @click.argument('spec_arg', default=None, required=False)
-@click.option('--spec', default=None, help=PipelineSpecClickType.help)
-@click.option('--pipeline-id', default=None,
-              help='id associated with the pipeline to be stopped')
+@click.option('--spec', default=None, type=PipelineSpecClickType(), help=PipelineSpecClickType.help)
+@click.option('--pipeline-id', default=None, type=PipelineIdClickType(),
+              help=PipelineIdClickType.help)
 @debug_option
 @profile_option
 @eat_exceptions
 @provide_api_client
 def delete_cli(api_client, spec_arg, spec, pipeline_id):
     """
-    Stops a delta pipeline and cleans up Databricks resources associated with it
+    Stops a delta pipeline and deletes its associated Databricks resources. The pipeline can be
+    resumed by deploying it again.
+
+    Usage:
+
+    databricks pipelines delete example.json
+
+    OR
+
+    databricks pipelines delete --spec example.json
+
+    OR
+
+    databricks pipelines delete --pipeline-id 1234
     """
-    # Only one out of spec/pipeline_id/spec_arg should be supplied
-    if bool(spec_arg) + bool(spec) + bool(pipeline_id) != 1:
-        raise RuntimeError('Either spec should be provided as an argument '
-                           'or option, or the pipeline-id should be provided')
-    if bool(spec_arg) or bool(spec):
-        src = spec_arg if bool(spec_arg) else spec
-        pipeline_id = _read_spec(src)["id"]
+    pipeline_id = _get_pipeline_id(spec_arg=spec_arg, spec=spec, pipeline_id=pipeline_id)
     PipelinesApi(api_client).delete(pipeline_id)
 
 
+@click.command(context_settings=CONTEXT_SETTINGS,
+               short_help='Gets a delta pipeline\'s current spec and status')
+@click.argument('spec_arg', default=None, required=False)
+@click.option('--spec', default=None, type=PipelineSpecClickType(), help=PipelineSpecClickType.help)
```

Comment:
It also gets the status. Arul and Michael and I debated a bunch when we first did this for delete, but we figured it was probably best to support all options there. I think we should do the same here, but check in with early customers and see if this is making it easier or harder to use.

Reply:
Classic REST will assign a unique ID on resource creation and then use that ID for subsequent requests. 

Accepting the entire spec for a delete() or similar call may seem harmless. But I have always believed that a good API has exactly one way to do a specific thing. Allowing to do it in multiple ways creates ambiguity, it can cause confusion, and it multiplies the amount of code that needs to be maintained (and kept backwards-compatible) in the future. 

One drawback of accepting a spec for get or delete etc. is that the user may think the entire spec must match, when in reality everything but ID is ignored. For example, a user may expect that the spec returned is equal to the spec passed in - that would not be the case. Or, on delete, the user may expect that it only deletes resources that match the whole spec, including, say, the name. But it will delete a pipeline with a matching ID, regardless of name. 

This kind of ambiguities can easily be avoided.